gcc: gcc_13 → gcc_14

fabianhjr opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

trofi committed:
vulkan-headers: build using ninja to fix `gcc-14` compatibility

When `vulkan-headers` detects `gcc-14` or above it enables c++ modules
support. But make-mased backend does not support it and fails as:

    $ nix build --impure --expr 'with import ./. {}; vulkan-headers.override {stdenv = gcc14Stdenv; }' -L

    CMake Error in CMakeLists.txt:
      The target named "Vulkan-Module" has C++ sources that may use modules, but
      modules are not supported by this generator:
        Unix Makefiles
      Modules are supported only by Ninja, Ninja Multi-Config, and Visual Studio
      generators for VS 17.4 and newer.  See the cmake-cxxmodules(7) manual for
      details.  Use the CMAKE_CXX_SCAN_FOR_MODULES variable to enable or disable
      scanning.

fabianhjr committed:
gcc: gcc_13 → gcc_14

fabianhjr committed:
treewide: add `-Wno-error=` due to compilation error with gcc_14

fabianhjr committed:
jam: broaden `-std=c89` due to compilation error with gcc_14

fabianhjr commented:
Cherry-picked a fix by @trofi from https://github.com/NixOS/nixpkgs/pull/340013

fabianhjr commented:
Currently doing a nixos-rebuild on x86_64-linux, stdenv + ~2k packages built so far (still missing another 1k builds for my config)

fabianhjr committed:
afdko: disable failing tests on gcc_14

fabianhjr committed:
treewide: add `-fpermissive` due to configuration error with gcc_14

vcunat commented:
https://gcc.gnu.org/gcc-14/porting_to.html

fabianhjr commented:
Tagging maintainers on the changes I'm not 100% sure on.

docbook2x: no maintainers listed
perlPackages.Po4a: no maintainers listed
afdko: @sternenseemann

Steam fhsenv drop of gconf is on a separated merge request and I intend to rebase this when that is merged and reaches current staging-next: https://github.com/NixOS/nixpkgs/pull/340633

alyssais commented:
Commit messages could use some more explanation — it's going to be very difficult to somebody looking at the history of a package in a couple of years to figure out why dependencies were dropped if it's not explained at all.

stigtsp commented:
We need to keep this in mind re the pelr 5.40.0 updated in https://github.com/NixOS/nixpkgs/pull/333286 as well

philiptaron commented on pkgs/applications/audio/mbrola/default.nix:
```diff
@@ -26,6 +26,10 @@ let
     # required for cross compilation
     makeFlags = [ "CC=${stdenv.cc.targetPrefix}cc" ];
 
+    env = lib.optionalAttrs stdenv.cc.isGNU {
+      NIX_CFLAGS_COMPILE = "-Wno-error=implicit-function-declaration";
+    };
```
As mentioned on Matrix by (I think) @emilazy, it would be great if there were tool assistance for notating parts of a derivation that should be attempted to be removed on next update.

@infinisil, maybe this is something that `nixpkgs-vet` could do in time... though upon further thought, it's closer to the job that ofBorg does.

philiptaron reviewed (changes requested):
Commit messages look great. Changes look great. A lot of things build (I made my computer sweat overnight on staging.)

Requesting changes for one more thing: a release note on the change in defaults.

fabianhjr commented:
Hi @vcunat, I believe most changes have been painless enough, I could help with further breakage during the next staging-next cycle.

fabianhjr commented:
@alyssais, removed the dependency drops from this PR, seems to build fine now (I think I carried over those drops by mistake while experimenting with this branch)

philiptaron commented:
Barring any objections, I'd like to land this in the next couple days.

fabianhjr committed:
zsh: add backported gcc_14 patch

Alpine
https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/70300

Buildroot
https://lore.kernel.org/all/20240720230500.26fbd225@windsurf/T/

OpenSUSE
https://build.opensuse.org/projects/home:jamborm:gcc14fixes/packages/zsh/files/zsh-4c89849c98172c951a9def3690e8647dae76308f.patch

OpenWRT
https://github.com/openwrt/packages/pull/24346/files

fabianhjr committed:
range-v3: use `-std=c++17` due to compilation error with gcc_13

fabianhjr commented on pkgs/development/libraries/mbedtls/generic.nix:
```diff
@@ -47,6 +47,10 @@ stdenv.mkDerivation rec {
   ];
 
+  env = lib.optionalAttrs stdenv.cc.isGNU {
+    NIX_CFLAGS_COMPILE = "-Wno-error=calloc-transposed-args";
```
This is a warning turned into an error back to not being an error; it would have been as dire on gcc 13 as with gcc 14.

/cc mbedtls nixpkgs maintainers: @raphaelr

raphaelr replied:
AFAIK this should only affect `mbedtls_2`, not `mbedtls` (v3). This issue has already been fixed for mbedtls v3: https://github.com/Mbed-TLS/mbedtls/pull/8751, but the fix wasn't backported to v2.

IMO the cleanest solution would be to apply that patch manually for v2, and leave v3 unchanged. It's unlikely that this patch will cause conflicts for v2 in the future since that branch only receives bugfixes/maintenance updates.
```nix
# pkgs/development/libraries/mbedtls/2.nix
{ callPackage
, fetchpatch
}:

callPackage ./generic.nix {
  version = "2.28.9";
  hash = "sha256-/Bm05CvS9t7WSh4qoMconCaD7frlmA/H9YDyJOuGuFE=";

  patches = [
    (fetchpatch {
      name = "gcc14-fix-calloc-argument-list.patch";
      url = "https://github.com/Mbed-TLS/mbedtls/commit/da2a33de0f2d0f7f179220d1a92f788d1459c3e3.patch";
      hash = "sha256-iKi1G0UDM2TlmVnSR8lRAkcQHnz7SQzJNQ3z2hAwZnw=";
    })
  ];
}
```

emilazy replied:
It's just a style nit, not sure it's worth patching over.

raphaelr replied:
I have no objections to the initial approach.

philiptaron replied:
Thanks for the discussion. I'll go with the team.

alyssais commented:
libaom no longer builds on aarch64-linux due to an internal compiler error, which I have [reported](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=116737) upstream.  That breaks 26810 packages.  Continuing to discuss and investigate in #staging:nixos.org.

alyssais commented:
We got lucky: https://github.com/NixOS/nixpkgs/pull/342306

philiptaron commented:
We've decided to revert this due to multiple internal compiler errors on aarch64.

- https://github.com/NixOS/nixpkgs/pull/340612

corngood commented:
a55b0ea0b170e31b2430ae9680eab435e2c5e9f1 seems to have broken jam.  See #349044.
